cpp/src/arrow/ipc/writer.cc

```diff
@@ -312,16 +314,18 @@ class RecordBatchSerializer : public ArrayVisitor {
 
   Status Visit(const BinaryArray& array) override { return VisitBinary(array); }
 
-  Status Visit(const ListArray& array) override {
+  template <typename ListArrayType, typename OffsetType>
```

Comment:
Perhaps the offset type should be available as `ListType::offset_type` and `LargeListType::offset_type`.